### DISCUSSION TITLE
An idea

### DISCUSSION THREAD

**Discussion by liming-600** [tjtanjin/QuickTax#2] [Q&A] [answered]
Could you please consider implementing this feature: redistributing taxes collected from high-income players to low-income players? I believe this could help improve balance in the game. It would be ideal to allow configuration of what constitutes low and high income within the plugin's settings.My English is not very good, I hope you can understand. Thank you very much for creating this plugin.

**Comment by tjtanjin** [comment]
Hey @clm123clm! Thank you for reaching out and excuse me for the long post ahead (I've consolidated quite some thoughts). Firstly, I am of the same opinion as you that being able to redistribute wealth from richer players to the less wealthy ones would be a nice addition. However, such a process can actually be rather complex which is why I'm slightly hesitant to directly support such a feature. Here are some scenarios to consider:

**Scenario 1:**
- Tax all players
- Redistribute all taxed money to the 50 lowest-income players

**Scenario 2:**
- Tax all players
- Redistribute all taxed money to bottom 10% of lowest-income players  

**Scenario 3:**
- Tax only the top 50 players
- Redistribute all taxed money to the 50 lowest-income players

**Scenario 4:**
- Tax only the top 50 players
- Redistribute all taxed money to the bottom 10% of lowest-income players

**Scenario 5:**
- Tax the top 10% of players
- Redistribute all taxed money to players with less than $1000

**Scenario 6:**
- Tax players with more than $1million
- Redistribute all taxed money to players with less than $1000

**Scenario 7:**
- Tax players with more than $1million
- Redistribute half the taxed money to players with less than $1000

**Scenario 8:**
- Tax players with more than $1million
- Redistribute all taxed money equally to all players

In the above scenarios, we see different ways of redistributing wealth (I'm sure there are more ways) and the trouble comes when we try to come up with a general configuration that supports all of these scenarios. It becomes quite apparent that there's no easy or clean way to set up a configuration for this - I'm personally not very much in favor of making the plugin configuration complex.

With all that said, I do believe it is possible to set out what you want to do. I can offer suggestions but I’ll need more context on which scenario you're looking at. It might be worth noting that the plugin currently supports the following which may be useful:

- Taxing players by balance (allows collecting taxes from high-income if you wish)
- Negative values in `bal-amount` (allows giving money to low-income players if you wish)
- A server balance (for you to consolidate all the taxes collected and thereby use the balance as you wish)

Note that it might be possible to use third-party plugins to achieve your goal as well since QuickTax also provides PAPI placeholders.

Of course, if you wish to have a further discussion, I'm more than happy to do so! Might be more convenient to continue on [discord](https://discord.gg/X8VSdZvBQY) if you have private use-cases to share or if you foresee a long conversation ahead. Hope this helps!

**Reply by liming-600** [reply]
Subject: Re: Your Response and Plugin Development

Dear [tjtanjin],

I'm delighted to receive your response, and I sincerely appreciate the time you took to address my queries. Reflecting on your suggestions, I can only imagine the challenges involved in developing plugins – it's indeed a laborious task, isn't it? Ha ha ha!

Regarding the implementation, I've given it some thought, and I agree that it's a daunting task. It might be wise to start with one of the more promising solutions, such as the "Scenario2" you mentioned. I believe it's a solid choice, although having the option to adjust percentages would be even better (considering there might be more than 10% of financially challenged players on the server, ha ha ha). It would be fantastic if administrators could configure this through a settings file, similar to other configurations. While having eight or more options might be ideal, having one viable and superior solution is certainly better than none. I wonder if you share this perspective?

As for developing third-party plugins to implement these functionalities, I too harbor a strong desire to create my own plugins. People who develop plugins are indeed cool, just like you, and I deeply admire your work. I've attempted to learn plugin development several times but have unfortunately met with little success. Initially, I tried to grasp the basics of Java through instructional videos on various platforms, but I only managed to scratch the surface before giving up. If it's not too much trouble, could you please recommend a step-by-step approach for learning plugin development? I would be more than eager to dive into it and would be immensely grateful for your guidance.

My English is not good.Forgive me for having to use translation software to communicate with you.Looking forward to your response.

Also, the country I am in makes it inconvenient for me to use Discord. I am very sorry.

[LiMing(github:clm123clm)]

**Reply by liming-600** [reply]
I noticed that you are a full stack developer and you are excellent. It is my honor to communicate with you!Your website is beautifully crafted. Also, you look really handsome!

**Reply by tjtanjin** [reply, marked as answer]
Hey @liming-600, thank you for the enthusiasm! While I agree that having a solution is better than having none at all, there isn't actually a superior solution amongst the scenarios I have suggested. The choice of scenario largely depends on the server and it's really more of a personal preference. In that regard, adopting just one scenario would be against the idea of keeping the plugin generic and flexible for all. What's more, if I cater to one scenario, there will come a time people arrive with request for other scenarios and it's a slippery slope to head down.


This brings me on to third-party plugins. Typically for larger servers, they have the resources to onboard developers to customize add-on features. If this is not possible, then it's up to server administrators to be creative and put together plugins to make things work. For example, looking for plugins that can grab the 10% lowest-income players and running commands on them.

For your desired scenario 2, the following 3 steps will need to be put together to work:
- Get 10% lowest-income players
- Get tax collected
- Run command to distribute money from tax against these list of low-income players

Is it possible to put together third-party plugins? Maybe, I do not know enough of the plugin market to be sure. However, if you want to go hands on and make your own plugin, this is achievable if you have some experiences working in java considering the lightweight logic involved as outlined in the 3 steps above.

For learning java, I personally picked it up first in a school project before subsequently doing various projects to gain exposure and experience. There are plenty of online resources these days and I apologize that I do not have the knowledge to point you to a good one. However, what I feel is important is practice. Don't just spend 12 hours a day reading tutorials/guides - actually get your hands dirty by working on a project to solidify your concepts and apply what you have learned.

All that said, Minecraft plugin development requires knowledge of not just java, but also managing dependencies, build tools and the often overlooked user experience. It'll be a bit hard to get into this without at least some basic knowledge of java (unless you already have a coding background). There's unfortunately no shortcut to this but it is a rewarding experience if you persevere on. Hope this helps you one way or another!

**Reply by liming-600** [reply]
Hello, Mr. Tanjin!Thank you very much for your response. Your words have made me deeply reflect on myself, and I realize that my thoughts were indeed somewhat selfish. I'm sorry for not considering your feelings.

I also lack sufficient understanding of the vast and relatively chaotic plugin market, not knowing how to combine third-party plugins. Regarding learning to create my own plugins, I will take your advice and persist in doing so.

I will try to add the desired features on my own when I have the ability in the future. Once again, thank you for taking the time to reply amidst your busy schedule. Your words have greatly benefited me, and it's my honor to communicate with you!

**Reply by tjtanjin** [reply]
You're too kind with your words :P Hope your server works out for you, feel free to reach out anytime!
